File: cpp/leetcode/137.cpp
public class Solution {
    int singleNumber(vector<int> &nums) {
        int ones = 0, twos = 0;
        for (int num : nums) {
            ones = (ones ^ num) & ~twos;
            twos = (twos ^ num) & ~ones;
        }
        return ones;
    }
}

public class Solution {
    int singleNumber(vector<int> &nums) {
        int count[32] = {0};
        int result = 0, n = nums.size();
        for (int i = 0; i < 32; i++) {
            for (int j = 0; j < n; j++) {
                if ((nums[j] >> i) & 1) count[i]++;
            }
            result |= ((count[i] % 3) << i);
        }
        return result;
    }
}

File: python/leetcode/137.py
class Solution:
    def singleNumber(self, nums):
        one, two = 0, 0
        for num in nums:
            # ones = ones ^ nums[i]; if (twos == 1) then ones = 0
            one = (one ^ num) & ~two
            # twos = twos ^ nums[i]; if (ones == 1) then twos = 0
            two = (two ^ num) & ~one
        return one
